File: src/main/java/hiiragi283/material/api/item/MaterialItemCasing.java
package hiiragi283.material.api.item;

import hiiragi283.material.api.machine.IMachineProperty;
import hiiragi283.material.api.machine.ModuleTrait;
import hiiragi283.material.api.material.HiiragiMaterial;
import hiiragi283.material.api.module.IModuleItem;
import hiiragi283.material.api.shape.HiiragiShapes;
import hiiragi283.material.util.CraftingBuilder;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.item.ItemStack;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.Set;

@ParametersAreNonnullByDefault
public class MaterialItemCasing extends MaterialItem implements IModuleItem {

    public MaterialItemCasing() {
        super(HiiragiShapes.CASING);
    }

    //    MaterialItem    //

    @Override
    protected void getRecipe(ITEM item, HiiragiMaterial material) {
        new CraftingBuilder(item.asItemStack(material))
                .setPattern("A A", "ABA", "A A")
                .setIngredient('A', HiiragiShapes.PLATE.getOreDict(material))
                .setIngredient('B', HiiragiShapes.FRAME.getOreDict(material))
                .build();
    }

    @Override
    protected void getModel(ITEM item) {
        HiiragiUtil.setModelSame(item.getObject());
    }

    //    IModuleItem    //

    @Override
    public int getProcessTime(ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .flatMap(HiiragiMaterial::getMachineProperty)
                .map(IMachineProperty::getProcessTime)
                .orElse(IModuleItem.super.getProcessTime(stack));
    }

    @Override
    public int getEnergyRate(ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .flatMap(HiiragiMaterial::getMachineProperty)
                .map(IMachineProperty::getEnergyRate)
                .orElse(IModuleItem.super.getEnergyRate(stack));
    }

    @Override
    public int getItemSlotCounts(ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .flatMap(HiiragiMaterial::getMachineProperty)
                .map(IMachineProperty::getItemSlotCounts)
                .orElse(IModuleItem.super.getItemSlotCounts(stack));
    }

    @Override
    public int getFluidSlotCounts(ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .flatMap(HiiragiMaterial::getMachineProperty)
                .map(IMachineProperty::getFluidSlotCounts)
                .orElse(IModuleItem.super.getFluidSlotCounts(stack));
    }

    @Override
    public Set<ModuleTrait> getModuleTraits(ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .flatMap(HiiragiMaterial::getMachineProperty)
                .map(IMachineProperty::getModuleTraits)
                .orElse(IModuleItem.super.getModuleTraits(stack));
    }

}

File: src/main/kotlin/hiiragi283/material/item/material/MaterialItemCasing.kt
package hiiragi283.material.item.material

import hiiragi283.material.api.item.MaterialItem
import hiiragi283.material.api.machine.MachinePropertyItem
import hiiragi283.material.api.machine.MachineTrait
import hiiragi283.material.api.material.HiiragiMaterial
import hiiragi283.material.init.HiiragiItems
import hiiragi283.material.init.HiiragiShapes
import hiiragi283.material.util.CraftingBuilder
import hiiragi283.material.util.itemStack
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.item.ItemStack
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

object MaterialItemCasing : MaterialItem(HiiragiShapes.CASING), MachinePropertyItem {

    override fun registerRecipe(material: HiiragiMaterial) {
        // 8x Plate + 1x Smithing Hammer -> 1x Casing
        if (HiiragiShapes.PLATE.canCreateMaterialItem(material)) {
            CraftingBuilder(itemStack(material))
                .setPattern("AAA", "ABA", "AAA")
                .setIngredient('A', HiiragiShapes.PLATE.getOreDict(material))
                .setIngredient('B', HiiragiItems.SMITHING_HAMMER, true)
                .build()
        }
        //Metal Former Recipe
        addMetalFormerRecipe(material, inputCount = 8)
        //Grinder Recipe
        addGrinderRecipe(material)
    }

    //    Client    //

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, worldIn: World?, tooltip: MutableList<String>, flagIn: ITooltipFlag) {
        this.addTooltip(stack, tooltip)
    }

    //    MachinePropertyItem    //

    override fun getProcessTime(stack: ItemStack): Int = getMaterial(stack)
        ?.machineProperty
        ?.processTime
        ?: super.getProcessTime(stack)

    override fun getEnergyRate(stack: ItemStack): Int = getMaterial(stack)
        ?.machineProperty
        ?.energyRate
        ?: super.getEnergyRate(stack)

    override fun getItemSlots(stack: ItemStack): Int = getMaterial(stack)
        ?.machineProperty
        ?.itemSlots
        ?: super.getItemSlots(stack)

    override fun getFluidSlots(stack: ItemStack): Int = getMaterial(stack)
        ?.machineProperty
        ?.fluidSlots
        ?: super.getFluidSlots(stack)

    override fun getMachineTraits(stack: ItemStack): Set<MachineTrait> = getMaterial(stack)
        ?.machineProperty
        ?.machineTraits
        ?: super.getMachineTraits(stack)

}